File: python/lmcc/strategy.py
"""Strategy: how a meaning travels (kernel §6), as data.

``{when?, requires?, visible?, fragments?, controls?, placement?,
routings?}`` or ``{choose: [{when, use}, …, {else}]}``. A routing is
``{from: "text" | "channel:<kind>", between | pattern | line_prefixed,
to: "@role" | "@role.<sub>", consume?}``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field as dc_field

from .errors import refuse

_KEYS = ("when", "requires", "visible", "fragments", "controls", "placement", "routings")
_PREDICATE_KEYS = ("capability", "not", "all", "any")
_TO = re.compile(r"^@role(\.[A-Za-z_][A-Za-z0-9_]*)?$")
_PLACEMENT = re.compile(r"^(controls\.[A-Za-z_][A-Za-z0-9_.]*|message:(system|user|assistant))$")
_FROM = re.compile(r"^(text|channel:[a-z_]+)$")
# outside the portable RE2 dialect (kernel §7a)
_NON_RE2 = re.compile(r"\(\?[=!>]|\(\?P?<|\\[1-9]|\\k<|[*+?}]\+")


@dataclass
class Strategy:
    when: dict | None = None
    requires: list[str] = dc_field(default_factory=list)
    visible: bool = True
    fragments: dict[str, str] = dc_field(default_factory=dict)
    controls: dict = dc_field(default_factory=dict)
    placement: dict[str, str] = dc_field(default_factory=dict)
    routings: list[dict] = dc_field(default_factory=list)
    choose: list[dict] | None = None      # [{"when": P, "use": Strategy}, {"else": Strategy}]

    # ------------------------------------------------------------ data

    def to_dict(self) -> dict:
        if self.choose is not None:
            out = []
            for alt in self.choose:
                if "else" in alt:
                    out.append({"else": alt["else"].to_dict()})
                else:
                    out.append({"when": dict(alt["when"]), "use": alt["use"].to_dict()})
            return {"choose": out}
        d: dict = {}
        if self.when is not None:
            d["when"] = dict(self.when)
        if self.requires:
            d["requires"] = list(self.requires)
        if not self.visible:
            d["visible"] = False
        if self.fragments:
            d["fragments"] = dict(self.fragments)
        if self.controls:
            d["controls"] = dict(self.controls)
        if self.placement:
            d["placement"] = dict(self.placement)
        if self.routings:
            d["routings"] = [dict(r) for r in self.routings]
        return d

    @classmethod
    def from_dict(cls, data: dict, *, where: str) -> "Strategy":
        if not isinstance(data, dict):
            refuse("entry-malformed", f"{where}: a strategy is an object")
        if "choose" in data:
            if set(data) != {"choose"} or not isinstance(data["choose"], list) or not data["choose"]:
                refuse("entry-malformed", f"{where}: choose is a non-empty list and stands alone")
            alts = []
            for i, alt in enumerate(data["choose"]):
                aw = f"{where}.choose[{i}]"
                if not isinstance(alt, dict):
                    refuse("entry-malformed", f"{aw}: an alternative is an object")
                if "else" in alt:
                    if set(alt) != {"else"} or i != len(data["choose"]) - 1:
                        refuse("entry-malformed", f"{aw}: else stands alone and comes last")
                    alts.append({"else": cls.from_dict(alt["else"], where=aw)})
                elif set(alt) == {"when", "use"}:
                    validate_predicate(alt["when"], where=f"{aw}.when")
                    alts.append({"when": alt["when"], "use": cls.from_dict(alt["use"], where=aw)})
                else:
                    refuse("entry-malformed", f"{aw}: an alternative is {{when, use}} or {{else}}")
            return cls(choose=alts)
        unknown = set(data) - set(_KEYS)
        if unknown:
            refuse("entry-malformed",
                   f"{where}: unknown strategy key(s) {sorted(unknown)}; known keys are {list(_KEYS)}")
        s = cls(when=data.get("when"), requires=list(data.get("requires", [])),
                visible=bool(data.get("visible", True)),
                fragments=dict(data.get("fragments", {})), controls=dict(data.get("controls", {})),
                placement=dict(data.get("placement", {})),
                routings=[dict(r) for r in data.get("routings", [])])
        s.validate(where=where)
        return s

    def validate(self, *, where: str) -> None:
        if self.choose is not None:
            return
        if self.when is not None:
            validate_predicate(self.when, where=f"{where}.when")
        for i, r in enumerate(self.routings):
            validate_routing(r, where=f"{where}.routings[{i}]")
        for target, place in self.placement.items():
            if not _TO.match(target) or not isinstance(place, str) or not _PLACEMENT.match(place):
                refuse("entry-malformed",
                       f"{where}.placement: {target!r}: {place!r} — a placement is "
                       f"'@role' or '@role.<sub>' → 'controls.<key>' or 'message:<role>'")
        for k, v in self.fragments.items():
            if k not in ("system", "user", "assistant") or not isinstance(v, str):
                refuse("entry-malformed", f"{where}.fragments: {k!r} must name a message role, text")
        if not self.visible and not self.routings and not self.placement:
            refuse("entry-malformed",
                   f"{where}: visible=false but no routing or placement serves the field — "
                   f"the value would be unrecoverable")

    # ------------------------------------------------------------ bind

    def select(self, capabilities: dict, *, role: str, name: str) -> "Strategy":
        """Resolve ``choose`` against the declared facts; check ``when`` and
        ``requires`` of the chosen alternative."""
        s = self
        while s.choose is not None:
            chosen = None
            for alt in s.choose:
                if "else" in alt or eval_predicate(alt["when"], capabilities):
                    chosen = alt.get("else") or alt["use"]
                    break
            if chosen is None:
                refuse("capability-missing",
                       f"role {role!r}: strategy {name!r}: no alternative of 'choose' "
                       f"holds for the declared capabilities and there is no else")
            s = chosen
        if s.when is not None and not eval_predicate(s.when, capabilities):
            refuse("capability-missing",
                   f"role {role!r}: strategy {name!r}: 'when' {s.when!r} is false for the "
                   f"declared capabilities")
        for fact in s.requires:
            if not capabilities.get(fact):
                refuse("capability-missing",
                       f"role {role!r}: strategy {name!r} requires capability {fact!r}, "
                       f"which the model does not declare")
        return s

    def bound(self, field_name: str) -> "Strategy":
        """A copy with ``{field}`` in fragments bound to the role's field."""
        return Strategy(when=self.when, requires=list(self.requires), visible=self.visible,
                        fragments={k: v.replace("{field}", field_name)
                                   for k, v in self.fragments.items()},
                        controls=dict(self.controls), placement=dict(self.placement),
                        routings=[dict(r) for r in self.routings])


def validate_routing(r: dict, *, where: str) -> None:
    if not isinstance(r, dict):
        refuse("entry-malformed", f"{where}: a routing is an object")
    src, to = r.get("from"), r.get("to")
    if not isinstance(src, str) or not _FROM.match(src):
        refuse("entry-malformed", f"{where}: 'from' is 'text' or 'channel:<part kind>'")
    if not isinstance(to, str) or not _TO.match(to):
        refuse("entry-malformed", f"{where}: 'to' is '@role' or '@role.<sub>'")
    unknown = set(r) - {"from", "to", "consume", "between", "pattern", "line_prefixed"}
    if unknown:
        refuse("entry-malformed", f"{where}: unknown routing key(s) {sorted(unknown)}")
    kinds = [k for k in ("between", "pattern", "line_prefixed") if k in r]
    if src == "text":
        if len(kinds) != 1:
            refuse("entry-malformed",
                   f"{where}: a text routing needs exactly one of between/pattern/line_prefixed")
        k = kinds[0]
        v = r[k]
        if k == "between":
            if not (isinstance(v, list) and len(v) == 2 and all(isinstance(x, str) and x for x in v)):
                refuse("entry-malformed", f"{where}: between is [open, close], non-empty strings")
        elif not isinstance(v, str) or not v:
            refuse("entry-malformed", f"{where}: {k} is a non-empty string")
        if k == "pattern":
            check_re2(v, where=where)
    elif kinds or r.get("consume"):
        refuse("entry-malformed", f"{where}: a channel routing takes no text extractor and no consume")


def check_re2(regex: str, *, where: str) -> None:
    unescaped = re.sub(r"\\[^1-9k]", "", regex)
    hit = _NON_RE2.search(unescaped)
    if hit:
        refuse("entry-malformed",
               f"{where}: regex {regex!r} uses {hit.group(0)!r}, which is outside the "
               f"portable RE2 dialect (no lookaround, backreferences, named groups, "
               f"atomic or possessive constructs)")
    try:
        re.compile(regex, re.DOTALL)
    except re.error as exc:
        refuse("entry-malformed", f"{where}: regex {regex!r} does not compile: {exc}")


def validate_predicate(p: object, *, where: str) -> None:
    if not isinstance(p, dict) or len(p) != 1:
        refuse("entry-malformed", f"{where}: a predicate is one of {_PREDICATE_KEYS}, one key")
    key, value = next(iter(p.items()))
    if key == "capability":
        if not isinstance(value, str):
            refuse("entry-malformed", f"{where}: 'capability' names a fact")
    elif key == "not":
        validate_predicate(value, where=where)
    elif key in ("all", "any"):
        if not isinstance(value, list):
            refuse("entry-malformed", f"{where}: {key!r} takes a list")
        for q in value:
            validate_predicate(q, where=where)
    else:
        refuse("entry-malformed", f"{where}: unknown predicate key {key!r}; known: {_PREDICATE_KEYS}")


def eval_predicate(p: dict, capabilities: dict) -> bool:
    key, value = next(iter(p.items()))
    if key == "capability":
        return bool(capabilities.get(value))
    if key == "not":
        return not eval_predicate(value, capabilities)
    if key == "all":
        return all(eval_predicate(q, capabilities) for q in value)
    return any(eval_predicate(q, capabilities) for q in value)

File: go/lmcc/strategy.go
package lmcc

import (
	"regexp"
	"strconv"
	"strings"
)

// Strategy: how a meaning travels (kernel §6), as data.
type Strategy struct {
	When      *Object // nil = none
	Requires  []string
	Visible   bool
	Fragments *Object // message role -> text
	Controls  *Object
	Placement *Object // "@role" | "@role.<sub>" -> "controls.<key>" | "message:<role>"
	Routings  []*Object
	Choose    []chooseAlt // non-nil: this strategy is a choose
}

type chooseAlt struct {
	When *Object // nil for else
	Use  *Strategy
}

var (
	strategyKeys  = []string{"when", "requires", "visible", "fragments", "controls", "placement", "routings"}
	predicateKeys = []string{"capability", "not", "all", "any"}
	toRE          = regexp.MustCompile(`^@role(\.[A-Za-z_][A-Za-z0-9_]*)?$`)
	placementRE   = regexp.MustCompile(`^(controls\.[A-Za-z_][A-Za-z0-9_.]*|message:(system|user|assistant))$`)
	fromRE        = regexp.MustCompile(`^(text|channel:[a-z_]+)$`)
	nonRE2        = regexp.MustCompile(`\(\?[=!>]|\(\?P?<|\\[1-9]|\\k<|[*+?}]\+`)
	escapes       = regexp.MustCompile(`\\[^1-9k]`)
)

func NewStrategy() *Strategy {
	return &Strategy{Fragments: NewObject(), Controls: NewObject(), Placement: NewObject(), Visible: true}
}

func (s *Strategy) ToJSON() *Object {
	if s.Choose != nil {
		list := []any{}
		for _, alt := range s.Choose {
			if alt.When == nil {
				list = append(list, Obj("else", alt.Use.ToJSON()))
			} else {
				list = append(list, Obj("when", DeepClone(alt.When), "use", alt.Use.ToJSON()))
			}
		}
		return Obj("choose", list)
	}
	out := NewObject()
	if s.When != nil {
		out.Set("when", DeepClone(s.When))
	}
	if len(s.Requires) > 0 {
		list := make([]any, len(s.Requires))
		for i, r := range s.Requires {
			list[i] = r
		}
		out.Set("requires", list)
	}
	if !s.Visible {
		out.Set("visible", false)
	}
	if s.Fragments.Len() > 0 {
		out.Set("fragments", s.Fragments.Clone())
	}
	if s.Controls.Len() > 0 {
		out.Set("controls", DeepClone(s.Controls))
	}
	if s.Placement.Len() > 0 {
		out.Set("placement", s.Placement.Clone())
	}
	if len(s.Routings) > 0 {
		list := make([]any, len(s.Routings))
		for i, r := range s.Routings {
			list[i] = DeepClone(r)
		}
		out.Set("routings", list)
	}
	return out
}

func strategyFromJSON(data *Object, where string) *Strategy {
	if data == nil {
		refusef("entry-malformed", "%s: a strategy is an object", where)
	}
	if data.Has("choose") {
		alts := data.List("choose")
		if data.Len() != 1 || len(alts) == 0 {
			refusef("entry-malformed", "%s: choose is a non-empty list and stands alone", where)
		}
		s := &Strategy{Choose: []chooseAlt{}}
		for i, raw := range alts {
			aw := where + ".choose[" + strconv.Itoa(i) + "]"
			alt, ok := raw.(*Object)
			if !ok {
				refusef("entry-malformed", "%s: an alternative is an object", aw)
			}
			if alt.Has("else") {
				if alt.Len() != 1 || i != len(alts)-1 {
					refusef("entry-malformed", "%s: else stands alone and comes last", aw)
				}
				s.Choose = append(s.Choose, chooseAlt{nil, strategyFromJSON(alt.Object("else"), aw)})
			} else if alt.Len() == 2 && alt.Has("when") && alt.Has("use") {
				validatePredicate(mustGet(alt, "when"), aw+".when")
				s.Choose = append(s.Choose, chooseAlt{alt.Object("when"), strategyFromJSON(alt.Object("use"), aw)})
			} else {
				refusef("entry-malformed", "%s: an alternative is {when, use} or {else}", aw)
			}
		}
		return s
	}
	for _, k := range data.Keys {
		if !contains(strategyKeys, k) {
			refusef("entry-malformed", "%s: unknown strategy key %q; known keys are %v", where, k, strategyKeys)
		}
	}
	s := NewStrategy()
	if data.Has("when") {
		s.When = data.Object("when")
		if s.When == nil {
			refusef("entry-malformed", "%s.when: a predicate is one of %v, one key", where, predicateKeys)
		}
	}
	for _, r := range data.List("requires") {
		rs, ok := r.(string)
		if !ok {
			refusef("entry-malformed", "%s: requires names facts", where)
		}
		s.Requires = append(s.Requires, rs)
	}
	s.Visible = data.Bool("visible", true)
	if f := data.Object("fragments"); f != nil {
		s.Fragments = f.Clone()
	}
	if c := data.Object("controls"); c != nil {
		s.Controls = c.Clone()
	}
	if p := data.Object("placement"); p != nil {
		s.Placement = p.Clone()
	}
	for _, r := range data.List("routings") {
		ro, ok := r.(*Object)
		if !ok {
			refusef("entry-malformed", "%s: each routing is an object", where)
		}
		s.Routings = append(s.Routings, ro.Clone())
	}
	s.validate(where)
	return s
}

func mustGet(o *Object, k string) any { v, _ := o.Get(k); return v }

func (s *Strategy) validate(where string) {
	if s.Choose != nil {
		return
	}
	if s.When != nil {
		validatePredicate(s.When, where+".when")
	}
	for i, r := range s.Routings {
		validateRouting(r, where+".routings["+strconv.Itoa(i)+"]")
	}
	for _, target := range s.Placement.Keys {
		place, ok := s.Placement.Str(target)
		if !toRE.MatchString(target) || !ok || !placementRE.MatchString(place) {
			refusef("entry-malformed", "%s.placement: %q: %v — a placement is '@role' or '@role.<sub>' → 'controls.<key>' or 'message:<role>'", where, target, mustGet(s.Placement, target))
		}
	}
	for _, k := range s.Fragments.Keys {
		if _, ok := s.Fragments.Str(k); !ok || (k != "system" && k != "user" && k != "assistant") {
			refusef("entry-malformed", "%s.fragments: %q must name a message role, text", where, k)
		}
	}
	if !s.Visible && len(s.Routings) == 0 && s.Placement.Len() == 0 {
		refusef("entry-malformed", "%s: visible=false but no routing or placement serves the field — the value would be unrecoverable", where)
	}
}

func validateRouting(r *Object, where string) {
	src, _ := r.Str("from")
	to, _ := r.Str("to")
	if !fromRE.MatchString(src) {
		refusef("entry-malformed", "%s: 'from' is 'text' or 'channel:<part kind>'", where)
	}
	if !toRE.MatchString(to) {
		refusef("entry-malformed", "%s: 'to' is '@role' or '@role.<sub>'", where)
	}
	for _, k := range r.Keys {
		if !contains([]string{"from", "to", "consume", "between", "pattern", "line_prefixed"}, k) {
			refusef("entry-malformed", "%s: unknown routing key %q", where, k)
		}
	}
	var kinds []string
	for _, k := range []string{"between", "pattern", "line_prefixed"} {
		if r.Has(k) {
			kinds = append(kinds, k)
		}
	}
	if src == "text" {
		if len(kinds) != 1 {
			refusef("entry-malformed", "%s: a text routing needs exactly one of between/pattern/line_prefixed", where)
		}
		switch kinds[0] {
		case "between":
			list := r.List("between")
			ok := len(list) == 2
			for _, x := range list {
				if s, isStr := x.(string); !isStr || s == "" {
					ok = false
				}
			}
			if !ok {
				refusef("entry-malformed", "%s: between is [open, close], non-empty strings", where)
			}
		case "pattern":
			re, ok := r.Str("pattern")
			if !ok || re == "" {
				refusef("entry-malformed", "%s: pattern is a non-empty string", where)
			}
			checkRE2(re, where)
		case "line_prefixed":
			if p, ok := r.Str("line_prefixed"); !ok || p == "" {
				refusef("entry-malformed", "%s: line_prefixed is a non-empty string", where)
			}
		}
	} else if len(kinds) > 0 || r.Bool("consume", false) {
		refusef("entry-malformed", "%s: a channel routing takes no text extractor and no consume", where)
	}
}

func checkRE2(re, where string) {
	if hit := nonRE2.FindString(escapes.ReplaceAllString(re, "")); hit != "" {
		refusef("entry-malformed", "%s: regex %q uses %q, which is outside the portable RE2 dialect (no lookaround, backreferences, named groups, atomic or possessive constructs)", where, re, hit)
	}
	if _, err := regexp.Compile("(?s)" + re); err != nil {
		refusef("entry-malformed", "%s: regex %q does not compile: %v", where, re, err)
	}
}

func validatePredicate(p any, where string) {
	po, ok := p.(*Object)
	if !ok || po.Len() != 1 {
		refusef("entry-malformed", "%s: a predicate is one of %v, one key", where, predicateKeys)
	}
	key := po.Keys[0]
	value, _ := po.Get(key)
	switch key {
	case "capability":
		if _, ok := value.(string); !ok {
			refusef("entry-malformed", "%s: 'capability' names a fact", where)
		}
	case "not":
		validatePredicate(value, where)
	case "all", "any":
		list, ok := value.([]any)
		if !ok {
			refusef("entry-malformed", "%s: %q takes a list", where, key)
		}
		for _, q := range list {
			validatePredicate(q, where)
		}
	default:
		refusef("entry-malformed", "%s: unknown predicate key %q; known: %v", where, key, predicateKeys)
	}
}

func evalPredicate(p *Object, capabilities *Object) bool {
	key := p.Keys[0]
	value, _ := p.Get(key)
	switch key {
	case "capability":
		return capabilities.Bool(value.(string), false)
	case "not":
		return !evalPredicate(value.(*Object), capabilities)
	case "all":
		for _, q := range value.([]any) {
			if !evalPredicate(q.(*Object), capabilities) {
				return false
			}
		}
		return true
	}
	for _, q := range value.([]any) {
		if evalPredicate(q.(*Object), capabilities) {
			return true
		}
	}
	return false
}

// selectFor resolves choose against the declared facts and checks when/requires.
func (s *Strategy) selectFor(capabilities *Object, role, name string) *Strategy {
	for s.Choose != nil {
		var chosen *Strategy
		for _, alt := range s.Choose {
			if alt.When == nil || evalPredicate(alt.When, capabilities) {
				chosen = alt.Use
				break
			}
		}
		if chosen == nil {
			refusef("capability-missing", "role %q: strategy %q: no alternative of 'choose' holds for the declared capabilities and there is no else", role, name)
		}
		s = chosen
	}
	if s.When != nil && !evalPredicate(s.When, capabilities) {
		refusef("capability-missing", "role %q: strategy %q: 'when' %s is false for the declared capabilities", role, name, MarshalJSON(s.When, -1))
	}
	for _, fact := range s.Requires {
		if !capabilities.Bool(fact, false) {
			refusef("capability-missing", "role %q: strategy %q requires capability %q, which the model does not declare", role, name, fact)
		}
	}
	return s
}

// bound returns a copy with {field} in fragments bound to the role's field.
func (s *Strategy) bound(fieldName string) *Strategy {
	out := &Strategy{When: s.When, Requires: append([]string{}, s.Requires...), Visible: s.Visible,
		Fragments: NewObject(), Controls: s.Controls.Clone(), Placement: s.Placement.Clone()}
	for _, r := range s.Routings {
		out.Routings = append(out.Routings, r.Clone())
	}
	for _, k := range s.Fragments.Keys {
		v, _ := s.Fragments.Str(k)
		out.Fragments.Set(k, strings.ReplaceAll(v, "{field}", fieldName))
	}
	return out
}
